### PR TITLE
Opaque type locations in error message for clarity.

### LINKED ISSUE #63167: Confusing error message with type unification of impl Traits
It took a colleague explaining to me that the return type of a function returning an impl Trait is unique across function definitions for me to understand the following error: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4ceda076c82b01ea279e8864518ab5d9

This becomes more important for async/await because of the automatic translation of `async fn foo() -> i64` to `fn foo() -> impl Future<Output = i64>`: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=b109446a49d22a632e395aacaa2cb8ff

In the first case, can the error maybe call out that the opaque types are different due to how the compiler treats them? And for the second error, can this be called out in async-specific language. That error happens when you forget to call `.await`.

### PR DESCRIPTION
**Pull request by gilescope**
Attempts to fix #63167

### CODE REVIEW

**Comment by cramertj** [comment]
This looks like a nice improvement, thanks!

@bors r+

File: src/librustc/infer/error_reporting/mod.rs [hunk 1]
```diff
@@ -1140,8 +1141,20 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                         &"type",
                         expected,
                         found,
-                        &format!(" ({})", values.expected.sort_string(self.tcx)),
-                        &format!(" ({})", values.found.sort_string(self.tcx)),
+                        &format!(" ({})",
+                             if let ty::Opaque(def_id, _) = values.expected.sty {
```
Comment (Centril) [inline src/librustc/infer/error_reporting/mod.rs]: The only difference between this and the one below is `values.expected` vs. `values.found`. It should be de-duplicated.

Comment (gilescope) [reply src/librustc/infer/error_reporting/mod.rs]: Thanks - I'd looked at adding a function but didn't quite seem worth it. A closure on the other had... looks much neater now.

**Pull request merged**